### PR TITLE
Garbage collection and bug fixes

### PR DESCRIPTION
**Pull request by SonicFrog**
Fixes issues when using agents from sieve and implements garbage collection for contagion.

Also adds proper subscribe mechanisms

### CODE REVIEW

**Review by tharvik — APPROVED** [review]

**Review by tharvik — APPROVED** [review]
> codecov/patch: 79.52% of diff hit (target 80.00%)

**Comment by SonicFrog** [comment]
Yeah should be fixed once I push the new tests

File: src/lib.rs [hunk 1]
```diff
@@ -461,6 +466,8 @@ where
 
                     let delivered = delivered.included().collect::<Vec<_>>();
 
+                    println!("{:?}", delivered);
```
Comment (tharvik) [inline src/lib.rs]: debug?

File: src/lib.rs [hunk 2]
```diff
@@ -730,17 +747,27 @@ pub mod test {
...
         if conflicts.len() == batch.info().size() {
             handle.deliver().await.expect_err("invalid delivery");
         } else {
-            let mut empty = FilteredBatch::new(batch.clone(), 0..length);
+            let mut empty = FilteredBatch::empty(batch.clone());
+
+            assert!(empty.is_empty());
 
             while let Ok(delivery) = handle.deliver().await {
+                println!("adding {} new sequences", delivery.len());
```
Comment (tharvik) [inline src/lib.rs]: debug?

File: src/lib.rs [hunk 3]
```diff
@@ -730,17 +747,27 @@ pub mod test {
...
                 empty.merge(delivery);
             }
 
+            println!("final delivery: {:?}", empty);
```
Comment (tharvik) [inline src/lib.rs]: debug?

File: src/lib.rs [hunk 4]
```diff
@@ -730,17 +747,27 @@ pub mod test {
...
         let batch = Arc::new(batch);
-        let length = batch.len();
 
         if conflicts.len() == batch.info().size() {
             handle.deliver().await.expect_err("invalid delivery");
         } else {
-            let mut empty = FilteredBatch::new(batch.clone(), 0..length);
+            let mut empty = FilteredBatch::empty(batch.clone());
+
+            assert!(empty.is_empty());
```
Comment (tharvik) [inline src/lib.rs]: look like you're testing the library outside of `cfg(test)`

**Review by tharvik — CHANGES_REQUESTED** [review]

**Review by tharvik — APPROVED** [review]

**Pull request merged**